File: gps_data_display/cpp/gps_data_display/src/gps_data_display_node.cpp
/**
 * @file gps_data_display_node.cpp
 * @brief ROS2 node for receiving and parsing GPS data from GPS driver
 *
 * This node subscribes to GPS data from the gps_ros2 driver and provides:
 * - Real-time GPS data display
 * - GPS status parsing and visualization
 * - Data logging to file
 * - Statistics calculation
 */

#include <memory>
#include <string>
#include <vector>
#include <fstream>
#include <iomanip>
#include <sstream>
#include <chrono>
#include <cmath>

#include "rclcpp/rclcpp.hpp"
#include "navigation_msgs/msg/gps_fix.hpp"

class GPSDataDisplayNode : public rclcpp::Node {
 public:
  GPSDataDisplayNode()
      : Node("gps_data_display_node"),
        total_messages_(0),
        valid_messages_(0),
        save_to_file_(false),
        log_interval_(1.0) {

    // Declare parameters
    this->declare_parameter("gps_topic", "gps/fix");
    this->declare_parameter("save_to_file", false);
    this->declare_parameter("log_file", "/tmp/gps_data.txt");
    this->declare_parameter("log_interval", 1.0);
    this->declare_parameter("show_raw_data", true);
    this->declare_parameter("show_status", true);

    // Get parameters
    gps_topic_ = this->get_parameter("gps_topic").as_string();
    save_to_file_ = this->get_parameter("save_to_file").as_bool();
    log_file_ = this->get_parameter("log_file").as_string();
    log_interval_ = this->get_parameter("log_interval").as_double();
    show_raw_data_ = this->get_parameter("show_raw_data").as_bool();
    show_status_ = this->get_parameter("show_status").as_bool();

    // Create subscriber
    subscription_ = this->create_subscription<navigation_msgs::msg::GpsFix>(
        gps_topic_, rclcpp::QoS(10),
        std::bind(&GPSDataDisplayNode::gps_callback, this, std::placeholders::_1));

    // Create statistics timer
    stats_timer_ = this->create_wall_timer(
        std::chrono::duration<double>(log_interval_),
        std::bind(&GPSDataDisplayNode::log_statistics, this));

    // Initialize statistics
    last_valid_time_ = this->now();
    min_lat_ = 90.0;
    max_lat_ = -90.0;
    min_lon_ = 180.0;
    max_lon_ = -180.0;
    max_speed_ = 0.0;
    max_sats_ = 0;

    // Open log file if enabled
    if (save_to_file_) {
      log_file_stream_.open(log_file_, std::ios::app);
      if (log_file_stream_.is_open()) {
        log_file_stream_ << "# GPS Data Log - Started at " << get_current_time_string() << std::endl;
        log_file_stream_ << "# timestamp,lat,lon,alt,status,sats,hdop,speed,heading,tx_ms" << std::endl;
        RCLCPP_INFO(this->get_logger(), "Log file opened: %s", log_file_.c_str());
      } else {
        RCLCPP_ERROR(this->get_logger(), "Failed to open log file: %s", log_file_.c_str());
        save_to_file_ = false;
      }
    }

    RCLCPP_INFO(this->get_logger(), "GPS Data Display Node initialized");
    RCLCPP_INFO(this->get_logger(), "  GPS topic: %s", gps_topic_.c_str());
    RCLCPP_INFO(this->get_logger(), "  Save to file: %s", save_to_file_ ? "true" : "false");
    if (save_to_file_) {
      RCLCPP_INFO(this->get_logger(), "  Log file: %s", log_file_.c_str());
    }
  }

  ~GPSDataDisplayNode() {
    if (log_file_stream_.is_open()) {
      log_file_stream_.close();
    }
  }

 private:
  void gps_callback(const navigation_msgs::msg::GpsFix::SharedPtr msg) {
    total_messages_++;

    // Parse GPS status
    std::string status_str = get_status_string(msg->status);
    bool is_valid = (msg->status > 0);

    if (is_valid) {
      valid_messages_++;
      last_valid_time_ = this->now();

      // Update statistics
      update_statistics(msg);

      // Update position bounds
      if (msg->latitude >= -90.0 && msg->latitude <= 90.0) {
        min_lat_ = std::min(min_lat_, msg->latitude);
        max_lat_ = std::max(max_lat_, msg->latitude);
      }
      if (msg->longitude >= -180.0 && msg->longitude <= 180.0) {
        min_lon_ = std::min(min_lon_, msg->longitude);
        max_lon_ = std::max(max_lon_, msg->longitude);
      }
    }

    // Display raw data
    if (show_raw_data_) {
      display_gps_data(msg, status_str);
    }

    // Show status summary
    if (show_status_) {
      display_status_summary(msg, status_str);
    }

    // Save to file if enabled
    if (save_to_file_ && log_file_stream_.is_open()) {
      save_to_log(msg);
    }
  }

  std::string get_status_string(int32_t status) {
    switch (status) {
      case 0: return "INVALID";
      case 1: return "SINGLE";
      case 2: return "DGPS/SBAS";
      case 4: return "RTK_FIXED";
      case 5: return "RTK_FLOAT";
      default: return "UNKNOWN(" + std::to_string(status) + ")";
    }
  }

  std::string get_status_color(int32_t status) {
    // Return ANSI color code based on status
    switch (status) {
      case 0: return "\033[31m";  // Red - Invalid
      case 1: return "\033[33m";  // Yellow - Single point
      case 2: return "\033[36m";  // Cyan - DGPS
      case 4: return "\033[32m";  // Green - RTK Fixed
      case 5: return "\033[34m";  // Blue - RTK Float
      default: return "\033[0m";  // Reset
    }
  }

  void display_gps_data(const navigation_msgs::msg::GpsFix::SharedPtr msg, const std::string& status_str) {
    std::string color = get_status_color(msg->status);
    std::string reset = "\033[0m";

    RCLCPP_INFO(this->get_logger(), "");
    RCLCPP_INFO(this->get_logger(), "========== GPS Data ==========");
    RCLCPP_INFO(this->get_logger(), "  Timestamp: %s", get_current_time_string().c_str());
    RCLCPP_INFO(this->get_logger(), "  Status: %s%s%s", color.c_str(), status_str.c_str(), reset.c_str());
    RCLCPP_INFO(this->get_logger(), "  Position:");
    RCLCPP_INFO(this->get_logger(), "    Latitude:  %.6f° %s", msg->latitude, msg->latitude >= 0 ? "N" : "S");
    RCLCPP_INFO(this->get_logger(), "    Longitude: %.6f° %s", msg->longitude, msg->longitude >= 0 ? "E" : "W");
    RCLCPP_INFO(this->get_logger(), "    Altitude:  %.3f m", msg->altitude);
    RCLCPP_INFO(this->get_logger(), "  Quality:");
    RCLCPP_INFO(this->get_logger(), "    Satellites: %d", msg->num_sats);
    RCLCPP_INFO(this->get_logger(), "    HDOP: %.2f", msg->hdop);
    RCLCPP_INFO(this->get_logger(), "  Motion:");
    RCLCPP_INFO(this->get_logger(), "    Speed:   %.3f m/s (%.1f km/h)", msg->speed, msg->speed * 3.6);
    RCLCPP_INFO(this->get_logger(), "    Heading: %.2f°", msg->heading);
    RCLCPP_INFO(this->get_logger(), "  Timing:");
    RCLCPP_INFO(this->get_logger(), "    GPS-PPS diff: %.3f ms", msg->tx);
    RCLCPP_INFO(this->get_logger(), "==============================");
  }

  void display_status_summary(const navigation_msgs::msg::GpsFix::SharedPtr msg, const std::string& status_str) {
    std::string color = get_status_color(msg->status);
    std::string reset = "\033[0m";

    // Single line summary for quick monitoring
    RCLCPP_INFO(this->get_logger(), "%s[GPS]%s %s | Sats:%d | HDOP:%.1f | %.6f,%.6f | Alt:%.1fm | Spd:%.1fkm/h | Head:%.1f°",
        color.c_str(), reset.c_str(),
        status_str.c_str(),
        msg->num_sats,
        msg->hdop,
        msg->latitude,
        msg->longitude,
        msg->altitude,
        msg->speed * 3.6,
        msg->heading);
  }

  void update_statistics(const navigation_msgs::msg::GpsFix::SharedPtr msg) {
    if (msg->speed > max_speed_) {
      max_speed_ = msg->speed;
    }
    if (msg->num_sats > max_sats_) {
      max_sats_ = msg->num_sats;
    }
  }

  void save_to_log(const navigation_msgs::msg::GpsFix::SharedPtr msg) {
    auto now = std::chrono::system_clock::now();
    auto now_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(now.time_since_epoch()).count();

    log_file_stream_ << std::fixed << std::setprecision(6)
                     << now_ns / 1e9 << ","
                     << msg->latitude << ","
                     << msg->longitude << ","
                     << msg->altitude << ","
                     << msg->status << ","
                     << msg->num_sats << ","
                     << msg->hdop << ","
                     << msg->speed << ","
                     << msg->heading << ","
                     << msg->tx
                     << std::endl;
  }

  void log_statistics() {
    if (total_messages_ == 0) {
      RCLCPP_INFO(this->get_logger(), "No GPS data received yet...");
      return;
    }

    double valid_rate = (total_messages_ > 0) ? (100.0 * valid_messages_ / total_messages_) : 0.0;

    RCLCPP_INFO(this->get_logger(), "");
    RCLCPP_INFO(this->get_logger(), "======= GPS Statistics =======");
    RCLCPP_INFO(this->get_logger(), "  Total messages: %lu", total_messages_);
    RCLCPP_INFO(this->get_logger(), "  Valid messages: %lu (%.1f%%)", valid_messages_, valid_rate);
    RCLCPP_INFO(this->get_logger(), "  Position bounds:");
    RCLCPP_INFO(this->get_logger(), "    Latitude:  [%.6f, %.6f]", min_lat_, max_lat_);
    RCLCPP_INFO(this->get_logger(), "    Longitude: [%.6f, %.6f]", min_lon_, max_lon_);
    RCLCPP_INFO(this->get_logger(), "  Max speed: %.3f m/s (%.1f km/h)", max_speed_, max_speed_ * 3.6);
    RCLCPP_INFO(this->get_logger(), "  Max satellites: %d", max_sats_);
    RCLCPP_INFO(this->get_logger(), "==============================");
  }

  std::string get_current_time_string() {
    auto now = std::chrono::system_clock::now();
    auto time = std::chrono::system_clock::to_time_t(now);
    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
        now.time_since_epoch()) % 1000;
    std::stringstream ss;
    ss << std::put_time(std::localtime(&time), "%Y-%m-%d %H:%M:%S");
    ss << '.' << std::setfill('0') << std::setw(3) << ms.count();
    return ss.str();
  }

  // ROS2 interfaces
  rclcpp::Subscription<navigation_msgs::msg::GpsFix>::SharedPtr subscription_;
  rclcpp::TimerBase::SharedPtr stats_timer_;

  // Parameters
  std::string gps_topic_;
  bool save_to_file_;
  std::string log_file_;
  double log_interval_;
  bool show_raw_data_;
  bool show_status_;

  // Log file stream
  std::ofstream log_file_stream_;

  // Statistics
  size_t total_messages_;
  size_t valid_messages_;
  rclcpp::Time last_valid_time_;

  // Position bounds
  double min_lat_, max_lat_;
  double min_lon_, max_lon_;
  double max_speed_;
  int max_sats_;
};

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<GPSDataDisplayNode>());
  rclcpp::shutdown();
  return 0;
}

File: gps_data_display/python/gps_data_display/gps_data_display_node.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
GPS Data Display Node (Python version)

This node subscribes to GPS data from the gps_ros2 driver and provides:
- Real-time GPS data display
- GPS status parsing and visualization
- Data logging to file
- Statistics calculation
"""

import os
from datetime import datetime
from typing import Optional

import rclpy
from rclpy.node import Node
from navigation_msgs.msg import GpsFix


class GPSDataDisplayNode(Node):
    """ROS2 Node for GPS data display"""

    # Status strings
    STATUS_STRINGS = {
        0: 'INVALID',
        1: 'SINGLE',
        2: 'DGPS/SBAS',
        4: 'RTK_FIXED',
        5: 'RTK_FLOAT',
    }

    # ANSI color codes
    STATUS_COLORS = {
        0: '\033[31m',  # Red - Invalid
        1: '\033[33m',  # Yellow - Single point
        2: '\033[36m',  # Cyan - DGPS
        4: '\033[32m',  # Green - RTK Fixed
        5: '\033[34m',  # Blue - RTK Float
    }
    RESET_COLOR = '\033[0m'

    def __init__(self):
        super().__init__('gps_data_display_node')

        # Statistics
        self.total_messages = 0
        self.valid_messages = 0

        # Position bounds
        self.min_lat = 90.0
        self.max_lat = -90.0
        self.min_lon = 180.0
        self.max_lon = -180.0
        self.max_speed = 0.0
        self.max_sats = 0

        # Log file
        self.log_file = None

        # Declare parameters
        self.declare_parameter('gps_topic', 'gps/fix')
        self.declare_parameter('save_to_file', False)
        self.declare_parameter('log_file', '/tmp/gps_data.txt')
        self.declare_parameter('log_interval', 1.0)
        self.declare_parameter('show_raw_data', True)
        self.declare_parameter('show_status', True)

        # Get parameters
        self.gps_topic = self.get_parameter('gps_topic').get_parameter_value().string_value
        self.save_to_file = self.get_parameter('save_to_file').get_parameter_value().bool_value
        self.log_file_path = self.get_parameter('log_file').get_parameter_value().string_value
        self.log_interval = self.get_parameter('log_interval').get_parameter_value().double_value
        self.show_raw_data = self.get_parameter('show_raw_data').get_parameter_value().bool_value
        self.show_status = self.get_parameter('show_status').get_parameter_value().bool_value

        # Create subscriber
        self.subscription = self.create_subscription(
            GpsFix,
            self.gps_topic,
            self.gps_callback,
            10
        )

        # Create statistics timer
        self.stats_timer = self.create_timer(
            self.log_interval,
            self.log_statistics
        )

        # Open log file if enabled
        if self.save_to_file:
            try:
                self.log_file = open(self.log_file_path, 'a')
                self.log_file.write(f'# GPS Data Log - Started at {self._get_current_time_string()}\n')
                self.log_file.write('# timestamp,lat,lon,alt,status,sats,hdop,speed,heading,tx_ms\n')
                self.get_logger().info(f'Log file opened: {self.log_file_path}')
            except Exception as e:
                self.get_logger().error(f'Failed to open log file: {e}')
                self.save_to_file = False

        self.get_logger().info('GPS Data Display Node initialized')
        self.get_logger().info(f'  GPS topic: {self.gps_topic}')
        self.get_logger().info(f'  Save to file: {self.save_to_file}')
        if self.save_to_file:
            self.get_logger().info(f'  Log file: {self.log_file_path}')

    def __del__(self):
        if self.log_file is not None:
            self.log_file.close()

    def gps_callback(self, msg):
        """GPS data callback"""
        self.total_messages += 1

        # Parse GPS status
        status_str = self._get_status_string(msg.status)
        is_valid = msg.status > 0

        if is_valid:
            self.valid_messages += 1

            # Update statistics
            self._update_statistics(msg)

            # Update position bounds
            if -90.0 <= msg.latitude <= 90.0:
                self.min_lat = min(self.min_lat, msg.latitude)
                self.max_lat = max(self.max_lat, msg.latitude)
            if -180.0 <= msg.longitude <= 180.0:
                self.min_lon = min(self.min_lon, msg.longitude)
                self.max_lon = max(self.max_lon, msg.longitude)

        # Display raw data
        if self.show_raw_data:
            self._display_gps_data(msg, status_str)

        # Show status summary
        if self.show_status:
            self._display_status_summary(msg, status_str)

        # Save to file if enabled
        if self.save_to_file and self.log_file is not None:
            self._save_to_log(msg)

    def _get_status_string(self, status):
        """Get status string from status code"""
        return self.STATUS_STRINGS.get(status, f'UNKNOWN({status})')

    def _get_status_color(self, status):
        """Get ANSI color code for status"""
        return self.STATUS_COLORS.get(status, self.RESET_COLOR)

    def _display_gps_data(self, msg, status_str):
        """Display GPS data"""
        color = self._get_status_color(msg.status)

        self.get_logger().info('')
        self.get_logger().info('========== GPS Data ==========')
        self.get_logger().info(f'  Timestamp: {self._get_current_time_string()}')
        self.get_logger().info(f'  Status: {color}{status_str}{self.RESET_COLOR}')
        self.get_logger().info('  Position:')
        lat_dir = 'N' if msg.latitude >= 0 else 'S'
        lon_dir = 'E' if msg.longitude >= 0 else 'W'
        self.get_logger().info(f'    Latitude:  {abs(msg.latitude):.6f}° {lat_dir}')
        self.get_logger().info(f'    Longitude: {abs(msg.longitude):.6f}° {lon_dir}')
        self.get_logger().info(f'    Altitude:  {msg.altitude:.3f} m')
        self.get_logger().info('  Quality:')
        self.get_logger().info(f'    Satellites: {msg.num_sats}')
        self.get_logger().info(f'    HDOP: {msg.hdop:.2f}')
        self.get_logger().info('  Motion:')
        self.get_logger().info(f'    Speed:   {msg.speed:.3f} m/s ({msg.speed * 3.6:.1f} km/h)')
        self.get_logger().info(f'    Heading: {msg.heading:.2f}°')
        self.get_logger().info('  Timing:')
        self.get_logger().info(f'    GPS-PPS diff: {msg.tx:.3f} ms')
        self.get_logger().info('==============================')

    def _display_status_summary(self, msg, status_str):
        """Display status summary"""
        color = self._get_status_color(msg.status)

        # Single line summary
        self.get_logger().info(
            f'{color}[GPS]{self.RESET_COLOR} {status_str} | '
            f'Sats:{msg.num_sats} | HDOP:{msg.hdop:.1f} | '
            f'{msg.latitude:.6f},{msg.longitude:.6f} | '
            f'Alt:{msg.altitude:.1f}m | Spd:{msg.speed * 3.6:.1f}km/h | Head:{msg.heading:.1f}°'
        )

    def _update_statistics(self, msg):
        """Update statistics"""
        if msg.speed > self.max_speed:
            self.max_speed = msg.speed
        if msg.num_sats > self.max_sats:
            self.max_sats = msg.num_sats

    def _save_to_log(self, msg):
        """Save GPS data to log file"""
        timestamp = datetime.now().timestamp()
        self.log_file.write(
            f'{timestamp:.6f},{msg.latitude},{msg.longitude},{msg.altitude},'
            f'{msg.status},{msg.num_sats},{msg.hdop},{msg.speed},'
            f'{msg.heading},{msg.tx}\n'
        )

    def log_statistics(self):
        """Log statistics"""
        if self.total_messages == 0:
            self.get_logger().info('No GPS data received yet...')
            return

        valid_rate = 100.0 * self.valid_messages / self.total_messages if self.total_messages > 0 else 0.0

        self.get_logger().info('')
        self.get_logger().info('======= GPS Statistics =======')
        self.get_logger().info(f'  Total messages: {self.total_messages}')
        self.get_logger().info(f'  Valid messages: {self.valid_messages} ({valid_rate:.1f}%)')
        self.get_logger().info('  Position bounds:')
        self.get_logger().info(f'    Latitude:  [{self.min_lat:.6f}, {self.max_lat:.6f}]')
        self.get_logger().info(f'    Longitude: [{self.min_lon:.6f}, {self.max_lon:.6f}]')
        self.get_logger().info(f'  Max speed: {self.max_speed:.3f} m/s ({self.max_speed * 3.6:.1f} km/h)')
        self.get_logger().info(f'  Max satellites: {self.max_sats}')
        self.get_logger().info('==============================')

    def _get_current_time_string(self):
        """Get current time string"""
        return datetime.now().strftime('%Y-%m-%d %H:%M:%S.%f')[:-3]


def main(args=None):
    rclpy.init(args=args)
    node = GPSDataDisplayNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()